proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
+
+The type of this instruction is:
+```
+[string] -> [i32, i32]
+
+Where:
+  * $alloc has a valid functype of [i32] -> [i32]
+```
+
+The instruction will execute the following steps:
+```
+  1. Let $length be the bytes required to encode the string with #encoding
+  2. Call $alloc with $length and receive an offset into #memidx
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `array.lift_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lift_memory #typeidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$base: i32, $length: i32] -> [$type]
+
+Where:
+  * $type is a type of the form (array $inner) taken from #typeidx
+  * #interface-instr* is valid with type [i32] -> [$inner]
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  2. Map #interface-instr* upon $offsets to yield the elements of the array
+```
+
+## `array.lower_memory`
+
+This is a structured instruction with the following syntax:
+```
+array.lower_memory #typeidx #memidx $alloc: #funcidx $width: #u32
+  interface-instr*
+end
+```
+
+The type of this instruction is:
+```
+[$type] -> [i32, i32]
+
+Where:
+  * $type is a valid type of the form (array $inner) taken from #typeidx
+  * $alloc has a valid functype of [i32] -> [i32]
+  * #interface-instr* is valid with type [i32, $inner] -> []
+```
+
+This instruction will execute the following steps:
+```
+  1. Let $elems be the elements of the array
+  2. Let $length be the length of the array
+  3. Let $bytes be '$length * $width'
+  4. Let $base = $alloc($bytes)
+  5. Let $offsets = take([$base, $base + $width, $base + $width * 2, ..], $length)
+  6. Let $params = zip($offsets, $elems)
+  7. Map #interface-instr* upon $params
+  8. Return [$base, $bytes]
+```
+
+## `invoke`
+
+The full syntax of this instruction is:
+```
```

Comment:
need two ways of calling an IT function: in the regular way and as a method with a designated receiver.

Reply:
Is it okay to add an invoke-method instruction later? I'm unsure how it would function as we don't have 'interface/service' or the notion of a 'receiver'.

Reply:
sure, we can add later. Not much later though :)